### PR TITLE
Scripting&Sound: Load sound from memory

### PR DESCRIPTION
**Pull request by the-maddin**
Fixing my own issue #3166 , this allows any LUA binary string containing valid wav data to be played as sound by the engine.

Such strings are created e.g. when communicating with the festival TTS server through LUAsocket.

This is accomplished by mirroring the functionality of ad.openAudioStream, and relevant parts of the sound system, to use a "fake" file in memory instead of a cfile as backend for FFmpegContext.

I've chosen to prefer audio streams over audio entries becuse they can be unloaded and do not clutter up the sound registry.

Test case: Having "L2-2_SN_02a.wav" (retail file) in your FS2 directory (not in any root), the following TBM will let Snipes greet you.
[tms-sct.tbm.txt](https://github.com/scp-fs2open/fs2open.github.com/files/6006767/tms-sct.tbm.txt)

### CODE REVIEW

File: code/sound/ffmpeg/FFmpegWaveFile.cpp [hunk 1]
```diff
@@ -236,6 +236,64 @@ bool FFmpegWaveFile::Open(const char* pszFilename, bool keep_ext)
...
 }
 
+// OpenMem
+bool FFmpegWaveFile::OpenMem(const uint8_t* snddata, size_t snd_len)
+{
+	using namespace libs::ffmpeg;
+
+
+	try {
+		m_ctx    = FFmpegContext::createContextMem(snddata, snd_len);
```
Comment (asarium) [inline code/sound/ffmpeg/FFmpegWaveFile.cpp:246]: Again duplicated code. The context creation is the only different thing and can be extracted.

**Review by asarium — CHANGES_REQUESTED** [review]

File: code/libs/ffmpeg/FFmpegContext.cpp [hunk 2]
```diff
@@ -135,6 +196,60 @@ std::unique_ptr<FFmpegContext> FFmpegContext::createContext(CFILE* mediaFile) {
...
+	}
+
+	auto avioBuffer = reinterpret_cast<uint8_t*>(av_malloc(AVIO_BUFFER_SIZE));
+
+	if (!avioBuffer) {
+		throw FFmpegException("Failed to allocate IO buffer!");
+	}
+
+	auto ioContext =
+		avio_alloc_context(avioBuffer, AVIO_BUFFER_SIZE, 0, &(instance->m_memsound), soundmem_read, nullptr, soundmem_seek);
```
[the remark this answers is not in the document]
Comment (the-maddin) [reply code/libs/ffmpeg/FFmpegContext.cpp:216]: To clarify, by "extract" you mean to pack off the functionality shared by the two versions into a new private version, and call that from both instead?

File: code/libs/ffmpeg/FFmpegContext.h [hunk 3]
```diff
@@ -15,14 +16,26 @@ namespace ffmpeg {
  * @brief Contains a FFmpeg context
  *
  * This should be used for making sure a FFmpeg context is always deallocated even if an exception is throws.
- * createContex can be used for opening a specific file for FFmpeg decoding
+ * createContext can be used for opening a specific file for FFmpeg decoding
  */
+
+struct MemSoundCursor {
```
[the remark this answers is not in the document]
Comment (the-maddin) [reply code/libs/ffmpeg/FFmpegContext.h:22]: Oops. I'll also rename it to fit the renamed functions.

File: code/scripting/api/libs/testing.cpp [hunk 4]
```diff
@@ -25,6 +31,55 @@ namespace api {
...
+	default:
+		LuaError(L, "Invalid audio stream type %d.", streamTypeEnum.index);
+		return ade_set_args(L, "o", l_AudioStream.Set(-1));
+	}
+	
+	
+	if (!snddata_arr.pushValue(L))
+	    return ade_set_args(L, "o", l_AudioStream.Set(-1));
+    if (!lua_isstring(L, -1)) {
+        lua_pop(L, 1);
```
[the remark this answers is not in the document]
Comment (the-maddin) [reply code/scripting/api/libs/testing.cpp:67]: Do you only mean the lua_isstring check or both? Failing to push something on the LUA stack is afaik an internal VM error, while passing not a string is indeed the scripter's error.

File: code/libs/ffmpeg/FFmpegContext.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (asarium) [reply code/libs/ffmpeg/FFmpegContext.cpp:216]: Yes, that is correct.

File: code/scripting/api/libs/testing.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (asarium) [reply code/scripting/api/libs/testing.cpp:67]: Yes, I only meant the type check.

File: code/libs/ffmpeg/FFmpegContext.cpp [hunk 5]
```diff
@@ -51,15 +53,70 @@ int64_t cfileSeek(void* ptr, int64_t offset, int whence) {
...
+            break;
+        case SEEK_CUR:
+            cursor_pos += offset;
+            break;
+        case SEEK_END:
+            cursor_pos = memsound->snddata.size() + offset;
+            break;
+    }
+
+    memsound->cursor_pos = MIN(static_cast<SCP_vector<uint8_t>::size_type>(cursor_pos), memsound->snddata.size());
```
Comment (asarium) [inline code/libs/ffmpeg/FFmpegContext.cpp:98]: Please use `std::min` to avoid using a macro.

File: code/libs/ffmpeg/FFmpegContext.cpp [hunk 6]
```diff
@@ -78,61 +135,76 @@ FFmpegContext::~FFmpegContext() {
...
+}
 
-		throw FFmpegException(SCP_string("Failed to get stream information! Error: ") + errorStr);
-	}
+std::unique_ptr<FFmpegContext> FFmpegContext::createContext(CFILE* mediaFile) {
+	Assertion(mediaFile != nullptr, "File pointer must be valid!");
+
+	std::unique_ptr<FFmpegContext> instance(new FFmpegContext(mediaFile));
+
+    return prepareInstance(std::move(instance), instance->m_file, cfileRead, cfileSeek);
```
Comment (asarium) [inline code/libs/ffmpeg/FFmpegContext.cpp:200]: Moving the pointer into the function suggests that ownership is transferred but the value is just returned again which is a bit misleading. Instead, you could also make `prepareInstance` just take a raw pointer and keep the unique pointer in this function.

**Review by asarium — CHANGES_REQUESTED** [review]
Just some minor stuff now. Looks good otherwise.

File: code/libs/ffmpeg/FFmpegContext.cpp (hunk 6 above)
Comment (the-maddin) [reply code/libs/ffmpeg/FFmpegContext.cpp]: Actually, looking at this again, `instance` shouldn't be passed at all. I'm making `prepareInstance` a nonstatic member, and call it _on_ `instance`

**Review by asarium — APPROVED** [review]

**Pull request merged**